pkg/controller/master/backup/restore.go

```diff
@@ -220,11 +273,117 @@ func (h *RestoreHandler) PersistentVolumeClaimOnChange(key string, pvc *corev1.P
 		return nil, nil
 	}
 
+	volumeName, err := getVolumeName(pvc)
+	if err != nil {
+		return nil, err
+	}
+
+	volume, err := h.volumeCache.Get(util.LonghornSystemNamespaceName, volumeName)
+	if err != nil {
+		return nil, nil
+	}
+
+	volumeCopy := volume.DeepCopy()
+	if volumeCopy.Annotations == nil {
+		volumeCopy.Annotations = make(map[string]string)
+	}
+
+	volumeCopy.Annotations[pvcNameSpaceAnnotation] = pvc.Namespace
+	volumeCopy.Annotations[pvcNameAnnotation] = pvc.Name
+	volumeCopy.Annotations[restoreNameAnnotation] = restoreName
+
+	if !reflect.DeepEqual(volume, volumeCopy) {
+		if _, err := h.volumes.Update(volumeCopy); err != nil {
+			return nil, err
+		}
+	}
+
 	logrus.Debugf("handling PVC updating %s/%s", pvc.Namespace, pvc.Name)
 	h.restoreController.EnqueueAfter(pvc.Namespace, restoreName, 5*time.Second)
 	return nil, nil
 }
 
+func (h *RestoreHandler) resolveLHEngineRef(lhEngine *lhv1beta2.Engine) (*lhv1beta2.Volume, error) {
+	refs := lhEngine.GetOwnerReferences()
+	if len(refs) == 0 {
+		return nil, fmt.Errorf("engine %s missing volume owner", lhEngine.Name)
+	}
+
+	ref := refs[0]
+	if ref.Kind != lhv1beta2.SchemeGroupVersion.WithKind("Volume").Kind {
+		return nil, fmt.Errorf("engine %s with wrong owner kind %v", lhEngine.Name, ref.Kind)
+	}
+
+	volume, err := h.volumeCache.Get(util.LonghornSystemNamespaceName, ref.Name)
+	if err != nil {
+		return nil, fmt.Errorf("engine %s owner volume %v not exist", lhEngine.Name, ref.Name)
+	}
+
+	if volume.UID != ref.UID {
+		return nil, fmt.Errorf("engine %s owner volume uuid not not consistent expected %v get %v",
+			lhEngine.Name, ref.UID, volume.UID)
+	}
+
+	return volume, nil
+}
+
+func (h *RestoreHandler) LHEngineOnChange(key string, lhEngine *lhv1beta2.Engine) (*lhv1beta2.Engine, error) {
+	if lhEngine == nil || lhEngine.DeletionTimestamp != nil || len(lhEngine.Status.RestoreStatus) == 0 {
+		return nil, nil
+	}
+
+	volume, err := h.resolveLHEngineRef(lhEngine)
+	if err != nil {
+		return nil, err
+	}
+
+	pvcNamespace, ok := volume.Annotations[pvcNameSpaceAnnotation]
+	if !ok {
+		return nil, nil
+	}
+
+	pvcName, ok := volume.Annotations[pvcNameAnnotation]
```

Comment:
Can we get pvc namespace and name from `volume.Status.KubernetesStatus.Namespace` and `volume.Status.KubernetesStatus.PvcName`?

Reply:
No, as https://github.com/harvester/harvester/pull/4869#discussion_r1431044095, `volume.Status.KubernetesStatus.Namespace` and `volume.Status.KubernetesStatus.PvcName` are filled after pvc already bound, however, during this time point, engine/replica phase already passed.